src/Main.hs

```diff
@@ -105,25 +107,63 @@ main = do
         contents <- readFile file'
         proceed contents (compile contents) $ \claims -> do
           let
-            handleRes (query, res) = case res of
-                  Unsat -> (True, typ <> " " <> prettyExp target <> " holds :)", show $ getSMT query)
-                  Sat -> (False, typ <> " " <> prettyExp target <> " does not hold :(", show $ getSMT query)
-                  SMT.Unknown -> (False, typ <> " " <> prettyExp target <> " could not be proved due to a solver timeout :(", show $ getSMT query)
-                  SMT.Error _ str -> (False, typ <> " " <> prettyExp target <> " could not be proved to due a solver error: " <> str, show $ getSMT query)
+            catModels results = [m | Sat m <- results]
```

Comment:
Maybe this is just me but I find this amount of nested `do`s, `let`s and `where`s to be quite difficult to follow. Many of these functions seem to be directly related to printing SMT stuff, and so maybe could be defined in `SMT.hs` instead?

Reply:
Tidied up a bit here, think it should be a little clearer now.

Reply:
Yes, much better!